### PR TITLE
Use setuptools_scm to automatically generate version number

### LINKED ISSUE #2197: Use setuptools_scm for generating a version number
[setuptools_scm](https://github.com/pypa/setuptools_scm/) can automatically generate version numbers based on git tags and versioning information. This will allow us to exactly identify which commit a user is reporting an issue with if it is not a released version. It will also remove one step of the release process.

### PR DESCRIPTION
**Pull request by ktbarrett**
Closes #2197.

### CODE REVIEW

**Comment by ktbarrett** [comment]
Fixed the versioning scheme so now it's generating the correct version.

File: pyproject.toml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [build-system]
-requires = ["setuptools", "wheel"]
+requires = ["setuptools", "wheel", "setuptools_scm"]
```
Comment (cmarqu) [inline pyproject.toml:2]: The example at https://pypi.org/project/setuptools-scm/ says "Note that the ``toml`` extra must be supplied.":
```
requires = ["setuptools>=42", "wheel", "setuptools_scm[toml]>=3.4"]
```
Maybe it works fine without it in most cases but maybe not all?

**Comment by eric-wieser** [comment]
I've no experience with setuptools_scm, but this looks reasonable assuming it works.

File: setup.py [hunk 2]
```diff
@@ -87,7 +84,11 @@ def package_files(directory):
 setup(
     name='cocotb',
     cmdclass={'build_ext': build_ext},
-    version=__version__,  # noqa: F821
+    use_scm_version={
+        'write_to': 'cocotb/_version.py',
+        'write_to_template': '__version__ = "{version}"',
```
Comment (eric-wieser) [inline setup.py:89]: nit: 
```suggestion
        'write_to_template': '__version__ = {version!r}',
```
although it only makes a difference if the version contains quotes or special characters.

File: setup.py [hunk 3]
```diff
@@ -87,7 +84,11 @@ def package_files(directory):
 setup(
     name='cocotb',
     cmdclass={'build_ext': build_ext},
-    version=__version__,  # noqa: F821
+    use_scm_version={
+        'write_to': 'cocotb/_version.py',
```
Comment (eric-wieser) [inline setup.py:88]: Stylistically, I sometimes like using
```suggestion
    use_scm_version=dict(
        write_to='cocotb/_version.py',
```
just to make the nested arguments resemble the non-nested ones.

File: pyproject.toml (hunk 1 above)
Comment (ktbarrett) [reply pyproject.toml:2]: That's if you are using pyproject.toml to configure the file.

File: setup.py [hunk 4]
```diff
@@ -87,7 +84,11 @@ def package_files(directory):
...
+        'write_to_template': '__version__ = "{version}"',
+        'version_scheme': 'release-branch-semver'
+    },
```
Comment (ktbarrett) [inline setup.py:91]:
```suggestion
    ),
```

**Comment by ktbarrett** [comment]
@eric-wieser The version is `cocotb-1.5.0.dev173+gdcafa990`. `173` is the number of commits on the current branch since the last release and `dcafa990` is the commit sha. On releases it should just be `cocotb-1.5.0`. The versioning scheme detects version branches (by checking if the branch name is the same as the last version tag) and will rev the patch version rather than the minor version in those cases.
Screenshot from 2020-11-16 07-36-09

**Comment by eric-wieser** [comment]
Since @imphil will likely be doing the release process, I'll leave it up to him to decide. Certainly this looks like an improvement for users installing from master, it just might add some slight friction at release time. Or it might just work.

**Comment by garmin-mjames** [comment]
Why is the version showing up as `v0.1.0.dev1+g5414f47` in the regressions?
https://github.com/cocotb/cocotb/pull/2198/checks?check_run_id=1406584885#step:22:103

**Comment by ktbarrett** [comment]
@garmin-mjames Shallow clone comes in without the tags. `setuptools_scm` uses git tags to determine the version number. We could probably add arguments to `action/checkout` step to have it retrieve the tags. I will make that an action.

**Comment by ktbarrett** [comment]
Added `fetch_depth: 0` to workflow to ensure tags are cloned. Fixed a typo causing failure.

**Comment by imphil** [comment]
Thanks for this change. I finally had time to experiment a bit with the versioning scheme (I know setuptools_scm from the past, and it works well itself). The tagging and branching detection works well, we just need to create a release branch right when we first do the RC release, otherwise we get slightly confusing version number (i.e. after tagging v1.5.0rc1, the next version number is v1.6.0.dev0 when we do everything on master). But in the worst case it's a slight "ugliness", but never a problem for actual releases.

**Review by imphil — APPROVED** [review]

**Pull request merged**